### ISSUE TITLE
Possibility to specifiy * in fetchedAttributes to synchronise all attributes of entries

### ISSUE THREAD

**Issue by coudot** [lsc-project/lsc#177]
For now we must define all attributes in fetchedAttributes, in source and destination.

An evolution would be to allow * value in fetchedAttributes to get all attributes of the entry and sync them.

**Comment by coudot** [comment]
Seems complex:  how to know the list of attributes?

**Comment by elecharny** [comment]
Indeed, I checked in 2.2 code base, we can already provide '*' as an attribute (as for '+' which will grab all the operation attributes):

```
...
        <fetchedAttributes>
          <string>*</string>
          <string>+</string>
...
```

returns all the attributes (user and operational).

**Comment by coudot** [comment]
We could maybe have a unit test on this ?

And an example in documentation

**Comment by elecharny** [comment]
I'll add both.

**Comment by davidcoutadeur** [comment]
First, you could check the behavior when you set * as fetchedAttributes:
* in source
* in destination

Especially, what is the behavior when * is specified in destination and the entry does not exist? No attribute is synchronized?

**Comment by elecharny** [comment]
The 'fetchedAttributes' parameter is only used when fetching an entry from either the source of destination LDAP servers, so I don't expect having any difference when using '*', except if it's used elsewhere ;-)

What could be problematic though is if the extra attributes are considered as added if they are missing from the destination, or deleted if absent from the source. 

We need more tests than what we have to cover all the bases:
- listed attributes from source and '*' from destination
- '*' from source and listed attributes from destination
- '*' from both

The semantic and expected result should be carefully stated

Last, not least, the current tests are problematic, I'm afraid. Most of them are using the *same* ldap server when it comes to sync a LDAP server with another. This has to be double checked.

**Comment by elecharny** [comment]
It was too good to be true...

using a '\*' in the source works just fine. Sadly using it in the destination does not work properly. The reason being the `BeanComparator.getWriteAttributes()` method uses the list of fetched attributes and iterate on it to create a new entry from the source (the ADD use case).

Well, technically speaking, adding a '\*' in the destination could mean 'inject whatever attribute present in the source', so we can simply check for the '*' value and go on.

**Comment by elecharny** [comment]
To add some clarity, let's say you use `*` for both the source and destination. Now we get an entry containing:

```
cn=CN0003,ou=ldap2ldap2TestTaskDst,ou=Test Data,dc=lsc-project,dc=org
description: Number three's descriptive text
sn: SN0003
cn: CN0003
objectclass: top
objectclass: inetOrgPerson
objectclass: person
objectclass: organizationalPerson

``` 
in the source and:

```
cn=CN0003,ou=ldap2ldap2TestTaskDst,ou=Test Data,dc=lsc-project,dc=org
sn: SN0003
cn: CN0003
objectclass: top
objectclass: inetOrgPerson
objectclass: person
objectclass: organizationalPerson

``` 
in the destination, then we have to determinate what to do with the missing `description` attribute in the destination.

The logic would be:
- if an attribute exists in source and in destination, then update values
- if an attribute exists in source and not in destination, then add the attribute in destination
- if an attribute exists in destination and not in source... ??? Should we ignore it or delete it?

When listing explicitly the attributes, we know what to do, because we have a list to process. Otherwise it's more complicated.
Currently, if we don't stipulate any attribute in the destination, then we build the list based on the sources attributes, and force, default/create attributes.

This is probably what we should do.

**Comment by elecharny** [comment]
Fixed in LSC-465

**Comment by elecharny** [comment]
After having discussed the various use cases for the destination service with David, here are the different treatments:

- If we have no fetched attributes: we use the list of source attributes to fetch the destination attributes
- If we have a list of attributes, we fetch them from the destination
- If we have '*' in the list, then we fetch all the user attributes from the destination
- if we have '+' in the list, then we fetch all the operational attributes from the destination
- If we have '*' and '+' in the list, then we fetch all the user and operational attributes from the destination
- If we have a list of attributes and '*', then we fetch all the  user attributes from the destination plus the listed attributes
- If we have a list of attributes and '+', then we fetch all the  operational attributes from the destination plus the listed attributes
- if we have a a list of attributes, '*' and '+', then we fetch all the  users and operational attributes from the destination  and we ignore the list of attributes

Last, not least: if the service is not a LDAP service, we will consider that '*' is just a standard column name, and we will go on.

**Comment by elecharny** [comment]
The fix is incorrect. When there is no entry destination, the entry should be created. The way it's handled with the fix, no entry get created, because we use '*' as an attribute, not a match.

**Comment by davidcoutadeur** [comment]
Hello,

After some thoughts, here is the proposition of meaning for '*' in **destination** service (and destination service only):

`*` as "all attributes in LDAP destination entry" has not much sense in the case of an entry creation. Indeed, in that case, the attributes considered for writing would result in an empty list.

Thus, I propose the following behavior: `*` in destination means:
- all attributes defined in source fetched attributes in LSC configuration
- plus all createValues, defaultValues, and forceValues in LSC configuration

I suggest adding a new boolean parameter for managing the deletion of extra attributes in destination entry not explicitly listed in destination fetchedAttributes. I will open another issue for that.

@elecharny @soisik any comment on that?

**Comment by davidcoutadeur** [comment]
Added corresponding documentation about * in fetchedAttributes here:

https://github.com/lsc-project/documentation/commit/ba524eff12adc14e867aa163d8f5e8eea02fad0a

**Comment by davidcoutadeur** [comment]
The PR seems ready to me. If no objection, I'll merge
